### PR TITLE
[Core][ModelPart] Geometry Container

### PR DESCRIPTION
**Pull request by tteschemacher**
Hi all,

this container will be a place to hold geometries within the model_part without any element or condition. This is essential for the CAD geometries but can be used for standard geometries, too.

I already collected a list of essential functions, however we might discuss also more functionalities. The data container shall be designed that all needed accesses can be achieved.

Thanks for your hints!

### CODE REVIEW

**Comment by loumalouomega** [comment]
Travis and appveyor are failing

**Comment by loumalouomega** [comment]
BTW, great job. I suggested this change some time ago, but I don't remember the reason (I think it was simply retro-compatibility)

File: kratos/includes/geometry_container.h [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+#include <sstream>
+#include <cstddef>
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "containers/pointer_vector_set.h"
+#include "containers/pointer_vector_map.h"
+#include "utilities/indexed_object.h"
```
Comment (loumalouomega) [inline kratos/includes/geometry_container.h]: Are all of these includes needed?

File: kratos/includes/geometry_container.h [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    void RemoveGeometry(GeometryType& ThisGeometry)
+    {
+        // HOW?
+        //mpGeometries->erase(ThisGeometry.Id());
+    }
+
+    /// Remove given element from geometry container
+    void RemoveGeometry(typename GeometryType::Pointer pThisGeometry)
+    {
+        // HOW?
```
Comment (loumalouomega) [inline kratos/includes/geometry_container.h]: Why these comments and commented code?

File: kratos/includes/model_part.h [hunk 3]
```diff
@@ -1555,6 +1561,8 @@ class KRATOS_API(KRATOS_CORE) ModelPart : public DataValueContainer, public Flag
 
     MeshesContainerType mMeshes; /// The container of all meshes
 
+    GeometryContainerType mGeometryContainer;
```
Comment (loumalouomega) [inline kratos/includes/model_part.h]: I would extend the API in a similar manner to the conditions and elements

**Review by loumalouomega — COMMENTED** [review]
Some comments for now

**Comment by philbucher** [comment]
@loumalouomega this is not ready to be merged, it serves as an initial point for discussion :)

**Comment by loumalouomega** [comment]
I know, I did my initial comments

File: kratos/includes/geometry_container.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/includes/geometry_container.h]: Actually, those are the parts which I think we should consider very much in our discussions to have something which will be working for everything.

**Comment by RiccardoRossi** [comment]
just to report one proposal:

about the proble of Ids, we consider two cases
- geometries with **user assigned Id** --> consider the Id as "positive" (or have a bit assigned to identifying this)
- geometries **without user-assigned Id** --> automatically generate one (for example a hash) but give it a negative sign (or control bit)

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee after long discussion achieves the first step for implementing the Geometries Container:
1. Geometry will be given an Id member
2. This Id will be filled in three ways: 
    1. Id is given by input file as a number and passed directly
    2. Id is given by a string and we compute a hash an set it as Id 
    3. If the geometry is constructed without any Id or name, then we will assign the this pointer as Id
3. The container would have an unordered_set of geometries using the Id of the geometry
4. The information about has given Id or given name or generated Id will be coded inside the Id itself
5. For now we don't add the mesh geometries to this container and only add the model geometry to it.

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee  also agrees on having the Geometry containers in sub modelparts with the same accumulative mechanism as elements and conditions

File: kratos/geometries/geometry.h [hunk 4]
```diff
@@ -144,6 +143,11 @@ class Geometry
     */
     typedef std::size_t SizeType;
 
+    /** Geometry shall behave similiar as indexed objects.
+    *   To allow query within data bases this type def is required.
+    */
+    typedef std::size_t result_type;
```
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: I would suggest you to use directly the `IndexType`

File: kratos/geometries/geometry.h [hunk 5]
```diff
@@ -555,26 +559,55 @@ class Geometry
...
-     // {
-     //     Geometry< Point >::PointsArrayType NewPoints;
+    ///@}
+    ///@name Id
+    ///@{
+
+    /// Id of this Geometry
+    IndexType Id() const
+    {
+        KRATOS_ERROR_IF(HasGeometryIdString())
```
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: I would not send an error even if the Id is a given hash

File: kratos/geometries/geometry.h [hunk 6]
```diff
@@ -555,26 +559,55 @@ class Geometry
...
+            << "Assigned geometry id is name of type string."
+            << std::endl;
+
+        return mId;
+    }
+
+    /// Id of this Geometry
+    IndexType GetHashId() const
```
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: This would GetId method, like the index object

File: kratos/geometries/geometry.h [hunk 7]
```diff
@@ -555,26 +559,55 @@ class Geometry
...
+    /// Sets Id with the use of the name of this geometry
+    void SetId(std::string name)
+    {
+        mId = GetNameHash(name);
+    }
 
-     //     //creating a geometry with the new points
-     //     Geometry< Point >::Pointer p_clone( new Geometry< Point >( NewPoints ) );
+    /// Gets the corresponding hash-Id to a string name
+    static inline IndexType GetNameHash(std::string name)
```
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: I would name this `CreateId` or `GenerateId`

File: kratos/geometries/geometry.h [hunk 8]
```diff
@@ -3189,20 +3192,53 @@ class Geometry
...
-    ///@name Static Member Variables
-    ///@{
-
-    // static const GeometryData msEmptyGeometryData;
-
-    ///@}
     ///@name Member Variables
     ///@{
 
+    IndexType mId;
```
Comment (pooyan-dadvand) [inline kratos/geometries/geometry.h]: The type should be `uint64_t` to be sure that it works also in 32bits machines.

**Review by pooyan-dadvand — CHANGES_REQUESTED** [review]
Start to getting shape but needs some minor changes

File: kratos/geometries/geometry.h (hunk 6 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: I changed it accordingly. Thx

File: kratos/geometries/geometry.h (hunk 8 above)
Comment (tteschemacher) [reply kratos/geometries/geometry.h]: I have some questions to this:
- do we need to change the indexed_object accordingly?
- does that mean that we need to update the IndexType accordingly, or start using a IdType?

**Comment by tteschemacher** [comment]
I am closing this PR as it is fully contained within #6166.

**Pull request closed without merging**